### PR TITLE
[MM-5998] move LaTeX support to a plugin

### PR DESCRIPTION
**Pull request by kosgrz**
#### Summary
This PR moves out LaTeX support from the webapp and extends the plugin framework to make it possible to implement LaTeX support.

Without a LaTeX plugin, it looks like this:

image

With the LaTeX plugin, it looks as follows:

image

#### LaTeX plugin
https://github.com/kosgrz/mattermost-plugin-latex

#### Ticket Link
  Fixes https://github.com/mattermost/mattermost-server/issues/11195

### CODE REVIEW

File: plugins/registry.js [hunk 1]
```diff
@@ -460,4 +460,10 @@ export default class PluginRegistry {
 
         return {id, showRHSPlugin: showRHSPlugin(id)};
     }
+
+    // Register a component to replace latex code with a component provided by the plugin.
```
Comment (lieut-data) [inline plugins/registry.js]: A high-level comment before diving in: I'm wondering if we might want to generalize this to support overriding arbitrary code blocks. Perhaps something like:
```
registerCodeBlockComponent(component, languages)
```

Where `languages` is an array -- in this case, `['tex', 'latex']. Then the component should expect to receive as a prop the content of the code block (perhaps `data-code` vs. forcing `data-latex`).

We could initially document this as only supporting the above two languages, but then in another effort, add support for arbitrary replacements without having to break the interface.

Comment (kosgrz) [reply plugins/registry.js]: @lieut-data this is definitely a good idea :+1:

**Comment by jasonblais** [comment]
Hey @kosgrz, you're latest rebase may have pulled additional changes you didn't intend to. Can you help take a look?

**Comment by kosgrz** [comment]
> Hey @kosgrz, you're latest rebase may have pulled additional changes you didn't intend to. Can you help take a look?

@jasonblais sorry, I didn't notice it before :/ . I will try to fix it today :+1:

**Comment by kosgrz** [comment]
@jasonblais sorry that I used force push, but didn't know how to fix it in another way :/

**Comment by jasonblais** [comment]
@kosgrz Thanks! That should be fine. It looks like the build failed, however. Might just be that you need to rebase your branch after the force push

**Comment by lieut-data** [comment]
@kosgrz, sorry for the delay on this one, and thank you for taking a stab at my initial proposed changes. This is still on my radar, and I'll try to give it some time this week.

File: components/post_markdown/index.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ function mapStateToProps(state, ownProps) {
     return {
         channel: getChannel(state, ownProps.channelId),
         pluginHooks: state.plugins.components.MessageWillFormat,
+        codeBlockPlugins: state.plugins.components.CodeBlockComponent || [],
```
Comment (lieut-data) [inline components/post_markdown/index.js]: nit: would it make sense to continue to use the term `codeBlockComponent`?

File: i18n/de.json [hunk 3]
```diff
@@ -2256,7 +2256,6 @@
   "invite_member.send2": "Einladungen senden",
   "invite_member.sending": " Sende",
   "invite_member.teamInviteLink": "Sie können außerdem Personen über den {link} einladen.",
-  "katex.error": "Konnte ihren Latex-Code nicht kompilieren. Bitte kontrollieren Sie die Syntax und versuchen es erneut.",
```
Comment (lieut-data) [inline i18n/de.json]: No need to modify anything but `en.json` -- the non-English files are handled automatically by the translator team, and this file will just get overwritten.

File: plugins/registry.js [hunk 4]
```diff
@@ -461,4 +461,23 @@ export default class PluginRegistry {
 
         return {id, showRHSPlugin: showRHSPlugin(id)};
     }
+
+    // Register a component to replace code blocks with components provided by plugins.
+    // Accepts a React component. Returns a unique identifier.
```
Comment (lieut-data) [inline plugins/registry.js]: Can you document the `languages` parameter here?

File: plugins/registry.js [hunk 5]
```diff
@@ -461,4 +461,23 @@ export default class PluginRegistry {
...
+    registerCodeBlockComponent(component, languages) {
```
Comment (lieut-data) [inline plugins/registry.js]: I'm wondering if the parameters should be reversed here, e.g. like `registerFilePreviewComponent`. If `languages` were a property of the component, I'd say this is fine, but I think we're overriding languages /with/ this component.

**Comment by aaronrothschild** [comment]
> One issue to resolve is the need to ship your plugin with existing Mattermost installations, otherwise exiting latex rendering will break, of course. @aaronrothschild, wondering how to proceed here. Complicating this further might be installations where plugins are currently disabled. Could be a good forcing function, actually!

This almost seems like it should be a system setting "Support Latex Processing" to make a smooth migration.

Just curious why did we originally want to take the processing of LaTeX out of the server?  That may help to explain to users why/how they'd migrate.

File: components/post_markdown/index.js (hunk 2 above)
Comment (kosgrz) [reply components/post_markdown/index.js]: Done :+1:

File: i18n/de.json (hunk 3 above)
Comment (kosgrz) [reply i18n/de.json]: Good to know :+1:

File: plugins/registry.js (hunk 4 above)
Comment (kosgrz) [reply plugins/registry.js]: Done :+1:

File: plugins/registry.js (hunk 5 above)
Comment (kosgrz) [reply plugins/registry.js]: I'm sorry, but it's not clear for me. How we can override languages here?

Comment (lieut-data) [reply plugins/registry.js]: Apologies, I'm wondering if we should just declare the signature of this API as:
```suggestion
    registerCodeBlockComponent(languages, component) {
```

to match how other functions arrange their parameters (e.g. `registerFilePreviewComponent`).

**Comment by lieut-data** [comment]
@aaronrothschild, this likely falls under the banner of feature flags -- uninstalling/disabling the latex plugin guarantees none of the corresponding code path can impact production, and lets us iterate on improvements independently.

> This almost seems like it should be a system setting "Support Latex Processing" to make a smooth migration.

I don't think we can smooth this anymore than:
* ship the latex plugin as pre-packaged
* release notes indicating that plugins and specifically this plugin must be enabled to preserve this functionality

I'm kind of inclined to think that's a good thing overall, given the desire to drive plugins. We just need to sort out the pre-packaged / plugin ownership issue first.

File: plugins/registry.js (hunk 5 above)
Comment (kosgrz) [reply plugins/registry.js]: Oh, I see now. Indeed, this is a good suggestion! Done :+1:.

**Comment by aaronrothschild** [comment]
Update: I've created Jira ticket https://mattermost.atlassian.net/browse/MM-18315 to productize this plugin and eventually make it available in the marketplace.

**Comment by cpoile** [comment]
@aaronrothschild Should we close this for now?

**Comment by aaronrothschild** [comment]
cc'ng @lieut-data @hanzei - We'll track this with the Jira ticket. 

When the plugin is done, we can merge the server code that removes Latex support from the server natively.

**Pull request closed without merging**